ntoskrnl/mm/amd64/init.c

```diff
@@ -207,7 +208,10 @@ MiInitializePageTable(VOID)
     ASSERT(__readcr4() & CR4_PGE);
 
     /* Enable no execute */
-    __writemsr(X86_MSR_EFER, __readmsr(X86_MSR_EFER) | EFER_NXE);
+    FeatureBits = KiGetFeatureBits();
+    if ((FeatureBits & KF_NX_BIT) == KF_NX_BIT)
```

Comment:
The AMD64 Architecture manual treats it as optional on section 5.6.3 https://support.amd.com/TechDocs/24593.pdf

Reply:
IMHO it should be the other way around: feature is unavailable unless clearly stated otherwise. Intel SDM, v.3A does not state nonconditional availability of this feature:

> Detect the presence of the execute disable bit capability using the CPUID instruction. CPUID.80000001H. EDX[bit 20] = 1 indicates the bit is available.
> If the bit is available and PAE is enabled, enable the execute disable bit capability by setting the IA32_EFER.NXE[bit 11] = 1. IA32_EFER is available if CPUID.80000001H.EDX[bit 20 or 29] = 1.
> If the execute disable bit capability is not available, a write to IA32_EFER.NXE produces a #GP exception.

So I think there is no harm in checking it.

Reply:
NX is required on x64 versions of Windows/ReactOS. It is already checked in KiInitializeCpu and the system will bugcheck if not available. KiInitializeCpu also enables NX in MSR_EFER already, so if can be removed entirely from this place.

Reply:
Ha, I missed that because it uses its own set of constants. I would argue EFER_xxx is the better naming scheme, so we should remove MSR_SCE, MSR_NXE et al, and move the EFER_xxx versions to NDK. Agreed?

Reply:
I'm not entirely sure, whether NDK is the right place to put these low level architecture specific things, a shared header seems right though. Maybe we create a (set of) header(s) for this stuff? In that case I would also vote for using structs/bitfields for all kinds of CPU related registers/structures rather than constants, since that is usually more convenient to use.